spoc/snipping/snipping_strategies.py

```diff
@@ -0,0 +1,240 @@
+            starts = np.random.randint(0, max_coord, size=number_to_choose)
+            ends = starts + length
+            chrom_frames.append(
+                pd.DataFrame({"chrom": chrom, "start": starts, "end": ends})
+            )
+        return pd.concat(chrom_frames)
+
+    def __repr__(self) -> str:
+        return f"<Triplet1DSnippingStrategy>"
+
+    @abstractmethod
+    def get_params():
+        raise NotImplementedError
+
+    @abstractmethod
+    def snip(
+        self,
+        pixels: Union[pd.DataFrame, PersistedPixels],
+        snip_positions: pd.DataFrame,
+        threads: int = 2,
+        override_snipping_value: bool = False,
+    ) -> pd.DataFrame:
+        """Do snipping"""
+        raise NotImplementedError
+
+
+class TripletCCT1DSnippingStrategy(SnippingStrategy):
+    """Implements snipping of 2D-regions based on a set of 1D-regions.
+    The 1D-regions are taken from the last column of pixels, which is assumed
+    to contain contacts on the second sister chromatid. The first two pixels
+    are assumed to contain contacts form the first sister chromatid."""
+
+    CIS_SNIPPING_QUERY = """
+        SELECT
+            t.position_id,
+            FLOOR((p.start_1 - t.pos)/{bin_size}::float) as offset_1,
+            FLOOR((p.start_2 - t.pos)/{bin_size}::float) as offset_2,
+            SUM(p.contact_count) as contacts
+        FROM {source_table} as p
+        INNER JOIN chunk as t ON t.chrom = p.chrom
+            and
+                abs(FLOOR((p.start_1 - t.pos)/{bin_size}::float))::int <= {pixel_offset}
+            and
+                abs(FLOOR((p.start_2 - t.pos)/{bin_size}::float))::int <= {pixel_offset}
+            and
+                abs(FLOOR((p.start_3 - (t.pos + {relative_offset}))/{bin_size}::float))::int <= {cis_selector_offset}
+        GROUP BY 1,2,3
+    """
+
+    def __init__(
+        self,
+        bin_size: int,
+        half_window_size: int,
+        snipping_value: SnippingValues,
+        **kwargs,
+    ):
+        """Override constructor to add additional parameters."""
+        super().__init__(bin_size, half_window_size, snipping_value, **kwargs)
+        self._position_slack: int = kwargs.get(
+            "position_slack", self._bin_size
+        )  # default is binsize
+        self._relative_offset: int = kwargs.get("relative_offset", 0)
+        self._expected = None
+
+    def get_params(self) -> dict:
+        return {
+            "bin_size": self._bin_size,
+            "half_window_size": self._half_window_size,
+            "snipping_value": self._snipping_value,
+            "position_slack": self._position_slack,
+            "relative_offset": self._relative_offset,
+        }
+
+    def _align_positions_to_bins(self, trans_positions: pd.DataFrame):
+        """Adds index and round positions to bins"""
+        if "pos" not in trans_positions.columns:
+            trans_positions = trans_positions.assign(
```

Comment:
Why do you use `assign` instead of `trans_positions['pos'] = ...`? I believe the latter is more readable.

Reply:
Ok, got it for here. We need to create a copy of the df to make changes on it.